### DISCUSSION TITLE
Floating element transforming into bottom sheet on mobile

### DISCUSSION THREAD

**Discussion by robinmetral** [floating-ui/floating-ui#1937] [React]
Hi!

I'm in the process of migrating a component from using Popper.js v2 to Floating UI. Very happy with the rewrite fo far, especially with the native TypeScript–thanks!

My component is basically an overflow menu that behaves like a popover on desktop viewports (attached to the reference element), but switches to a bottom sheet on mobile viewports:

| Desktop | Mobile |
| --- | --- |
| <img width="1357" alt="Popover on a wide viewport" src="https://user-images.githubusercontent.com/35560568/178956423-897c8db7-edcf-4f60-893d-c9f8f345e086.png"> | <img width="320" alt="Bottom sheet on a narrow viewport" src="https://user-images.githubusercontent.com/35560568/178956542-6ad08d45-fd5c-48cc-8c78-03ddd26ea04b.png"> |

With Popper.js, I had a custom modifier that looked something like this:

```tsx
const mobilePosition: Modifier<'mobilePosition', { state: State }> = () => ({
  name: 'mobilePosition',
  enabled: true,
  phase: 'write',
  fn({ state }) {
    const isMobile = window.matchMedia('(max-width: 600px)').matches;
    if (isMobile) {
      state.styles.popper = {
        position: 'fixed',
        width: '100%',
        bottom: 0,
        left: 0,
        right: 0,
      };
    };
  },
});
```

I tried a similar approach with the Floating UI [size middleware](https://floating-ui.com/docs/size)...

```tsx
const mobilePosition = size({
  apply({ elements }) {
    const isMobile = window.matchMedia('(max-width: 600px)').matches;
    if (isMobile) {
      elements.floating.style = {
        position: 'fixed',
        width: '100%',
        bottom: 0,
        left: 0,
        right: 0,
      };
    }
  },
});
```

...but this didn't quite work. It looks like this clashes with the positioning styles passed inline to the tooltip element (`top`, `left` and `position` as recommended in docs). I imagine that the lifecycle looks slightly different in Floating UI so this strategy doesn't work here?

As a workaround I found two possible solutions:

1. don't pass `top` and `left` to the floating element as `style` at all, and handle both the desktop and mobile scenarios in the `size` middleware 👉 [CodeSandbox](https://codesandbox.io/s/floating-ui-bottom-sheet-with-size-middleware-nwzr5n)
2. don't use the `size` middleware at all, and handle the mobile scenario inline in the JSX 👉 [CodeSandbox](https://codesandbox.io/s/floating-ui-bottom-sheet-inline-dur1ul)

Two questions:
- is one of the two approaches above recommended when overriding positioning from Floating UI? Performance implications that I'm not aware of?
- are you interested in me contributing docs for this use case? I could see the pattern (basically stretch and stick on mobile) being of interest for others

Let me know, and thanks a lot!

**Comment by Justinidlerz** [comment]
I tried solution 1 and upgraded to the latest floating-ui (0.26.x), but it is not working now.

**Reply by Justinidlerz** [reply]
It is shown as that the apply method will be called while the window resizing on desktop size, but the mobile size will not

**Reply by Justinidlerz** [reply]
Sorry, It seems that the API of floating-ui has been updated, I changed the implementation for solution 1 and it also works.
```tsx
import "./styles.css";
import { autoUpdate, useFloating, size } from "@floating-ui/react-dom";

const mobileStyles = {
  bottom: 0,
  left: 0,
  right: 0,
  top: "auto",
  width: "100%",
};

export default function App() {
  const { strategy, refs } = useFloating({
    placement: "bottom",
    strategy: "fixed",
    whileElementsMounted: autoUpdate,
    middleware: [
      size({
        apply({ elements, x, y }) {
          const isMobile = window.matchMedia("(max-width: 600px)").matches;
          console.log(isMobile);
          Object.assign(elements.floating.style, {
            top: `${y}px`,
            left: `${x}px`,
            bottom: "auto",
            right: "auto",
            width: "auto",
            ...(isMobile && { ...mobileStyles }),
          });
        },
      }),
    ],
  });
  return (
    <div className="app">
      <button ref={refs.setReference}>Reference</button>
      <div
        className="floating"
        ref={refs.setFloating}
        style={{ position: strategy }}
      >
        I am floating
      </div>
    </div>
  );
}

```
